File: src/main/java/ca/yyx/hu/view/ProjectionView.java
package ca.yyx.hu.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import ca.yyx.hu.App;
import ca.yyx.hu.decoder.VideoDecoder;
import ca.yyx.hu.utils.AppLog;

/**
 * @author algavris
 * @date 09/11/2016.
 */

public class ProjectionView extends SurfaceView implements SurfaceHolder.Callback {
    protected VideoDecoder mVideoDecoder;
    private  SurfaceHolder.Callback mSurfaceCallback;


    public ProjectionView(Context context) {
        super(context);
        init();
    }

    public ProjectionView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public ProjectionView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void setSurfaceCallback(SurfaceHolder.Callback surfaceCallback) {
        mSurfaceCallback = surfaceCallback;
    }

    private void init() {
        mVideoDecoder = App.get(getContext()).videoDecoder();
        getHolder().addCallback(this);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        mVideoDecoder.stop("onDetachedFromWindow");
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        AppLog.i("holder " + holder);
        if (mSurfaceCallback != null) {
            mSurfaceCallback.surfaceCreated(holder);
        }
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        AppLog.i("holder %s, format: %d, width: %d, height: %d", holder, format, width, height);
        mVideoDecoder.onSurfaceHolderAvailable(holder, width, height);
        if (mSurfaceCallback != null) {
            mSurfaceCallback.surfaceChanged(holder, format, width, height);
        }
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        AppLog.i("holder " + holder);
        mVideoDecoder.stop("surfaceDestroyed");
        if (mSurfaceCallback != null) {
            mSurfaceCallback.surfaceDestroyed(holder);
        }
    }
}

File: src/main/java/ca/yyx/hu/view/ProjectionView.kt
package ca.yyx.hu.view

import android.content.Context
import android.util.AttributeSet
import android.view.SurfaceHolder
import android.view.SurfaceView

import ca.yyx.hu.App
import ca.yyx.hu.decoder.VideoDecoder
import ca.yyx.hu.utils.AppLog

/**
 * @author algavris
 * *
 * @date 09/11/2016.
 */

class ProjectionView : SurfaceView, SurfaceHolder.Callback {
    private var mVideoDecoder: VideoDecoder? = null
    private var mSurfaceCallback: SurfaceHolder.Callback? = null


    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    fun setSurfaceCallback(surfaceCallback: SurfaceHolder.Callback) {
        mSurfaceCallback = surfaceCallback
    }

    private fun init() {
        mVideoDecoder = App.get(context).videoDecoder()
        holder.addCallback(this)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        mVideoDecoder?.stop("onDetachedFromWindow")
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        AppLog.i("holder " + holder)
        mSurfaceCallback?.surfaceCreated(holder)
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        AppLog.i("holder %s, format: %d, width: %d, height: %d", holder, format, width, height)
        mVideoDecoder?.onSurfaceHolderAvailable(holder, width, height)
        mSurfaceCallback?.surfaceChanged(holder, format, width, height)
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        AppLog.i("holder " + holder)
        mVideoDecoder?.stop("surfaceDestroyed")
        mSurfaceCallback?.surfaceDestroyed(holder)
    }
}
